pkg/metrics/metric/metric.go

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metric
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// WithMetadata is the interface implemented by any metric defined in this package. These typically embed existing
+// prometheus metric types and add additional metadata to the definitions which can be used to generate proper metric
+// documentation from de definitions. In addition, these metrics have the concept of being enabled or disabled which
+// is used in place of conditional registration so all metric types can always be registered for the purposes of
+// documenting them.
+type WithMetadata interface {
+	IsEnabled() bool
+	SetEnabled(bool)
+	Opts() Opts
+}
+
+// metric is a "base" structure which can be embedded to provide common functionality.
+type metric struct {
+	enabled bool
+	opts    Opts
+}
+
+func (b *metric) IsEnabled() bool {
+	return b.enabled
+}
+
+func (b *metric) SetEnabled(e bool) {
+	b.enabled = e
+}
+
+// Vec is a generic type to describe the vectorized version of another metric type, for example Vec[Counter] would be
+// our version of a prometheus.CounterVec.
+type Vec[T any] interface {
+	prometheus.Collector
+	WithMetadata
+
+	CurryWith(labels prometheus.Labels) (Vec[T], error)
+	GetMetricWith(labels prometheus.Labels) (T, error)
+	GetMetricWithLabelValues(lvs ...string) (T, error)
+	With(labels prometheus.Labels) T
+	WithLabelValues(lvs ...string) T
```

Comment:
I added the comments to the interface and modified it to fit with the generic interfaces. I think this will end up neater in code completion and generated docs.